src/compute-engine/boxed-expression/match.ts

```diff
@@ -352,82 +403,217 @@ function matchArguments(
-        }
-
-        // Determine the value to return for the wildcard
-        let value: BoxedExpression;
-        if (j < 1) {
-          // A standard/non-optional Sequence Wildcard with no match...
-          if (!argName.startsWith('___') && j < 1) return null;
-          // Otherwise must be an optional-sequence match: this is permitted.
-          if (expr.operator === 'Add') value = ce.Zero;
-          else if (expr.operator === 'Multiply') value = ce.One;
-          else value = ce.Nothing;
-        } else if (j === 1) {
-          // Capturing a single element/operand
-          value = ops.shift()!;
-        } else {
-          // >1 operands captured
-          const def = ce.lookupDefinition(expr.operator);
-          const args = ops.splice(0, j);
-          if (def && isOperatorDef(def) && def.operator.associative) {
-            value = ce.function(expr.operator, args, { canonical: false });
+  return matchRemaining(patterns, substitution);
+
+  /*
+   * Local f().
+   */
+  /**
+   * Match a list of patterns against *remaining* (locally scoped) `ops` and consume this list along
+   * the way, whilst appending to `substitution`. If a complete/successful match, return this new
+   * substitution; else return `null`.
+   *
+   * @note: calls recursively (permutations) for sequence wildcards
+   *
+   * @param patterns
+   * @param substitution
+   * @returns
+   */
+  function matchRemaining(
+    patterns: ReadonlyArray<BoxedExpression>,
+    substitution: BoxedSubstitution
+  ): BoxedSubstitution | null {
+    let result: BoxedSubstitution | null = { ...substitution };
+
+    let i = 0; // Index in pattern
+
+    while (i < patterns.length) {
+      const pat = patterns[i];
+      const argName = wildcardName(pat);
+
+      if (argName !== null) {
+        if (argName.startsWith('__')) {
+          // Match 1 or more expressions (__) or 0 or more (___)
+          const nextPattern = patterns[i + 1];
+
+          const isOptionalSeq = argName.startsWith('___');
+
+          if (nextPattern === undefined) {
+            // No more args in the pattern after, go till the end
+            if (ops.length === 0 && !isOptionalSeq) return null;
+            result = captureWildcard(argName, captureOps(ops.length), result);
           } else {
-            value = ce.function('Sequence', args, { canonical: false });
+            /** Total qty. of operands to be consumed by this sequence (of those remaining).
+             * If a non-optional sequence, minimum must be 1.
+             */
+            let j = isOptionalSeq ? 0 : 1;
+
+            // The next pattern should not be another required sequence wildcard
+            // (^@todo?: validate beforehand; should never be permitted?)
+            console.assert(
+              !(
+                isWildcard(nextPattern) &&
+                wildcardType(nextPattern) === 'Sequence'
+              )
+            );
```

Comment:
'Validation' step as mentioned (for call to `expr.match()`) would still be needed to verify this